File: tools/project/templates/cpp/app_core_fragment.cpp
// Tellusim Fragment Application Template

#include <TellusimApp.h>
#include <core/TellusimCore.h>
#include <math/TellusimMath.h>
#include <platform/TellusimPlatforms.h>

/*
 */
using namespace Tellusim;

/*
 */
class Application {
		
	public:
		
		// constructor initializing the application with command-line arguments
		Application(int32_t argc, char **argv) : app(argc, argv) {
			
		}
		
		// destructor cleans up resources
		virtual ~Application() {
			
			// finish context
			if(window) window.finish();
		}
		
		// create function responsible for initialization
		virtual bool create() {
			
			// create app
			if(!app.create()) {
				TS_LOG(Error, "Application::create(): can't create App\n");
				return false;
			}
			
			// create window
			if(!create_window()) return false;
			
			// create resources
			if(!create_resources()) return false;
			
			// additional initialization code can go here
			
			return true;
		}
		
		// update function for logic updates
		virtual bool update() {
			
			// your update logic here
			
			return true;
		}
		
		// render function to handle the drawing commands
		virtual bool render(Command &command) {
			
			// fragment uniform parameters
			struct FragmentParameters {
				float32_t aspect;
				float32_t time;
			};
			
			FragmentParameters parameters;
			parameters.aspect = (float32_t)window.getWidth() / (float32_t)window.getHeight();
			parameters.time = (float32_t)Time::seconds();
			
			// draw fullscreen triangle
			command.setPipeline(pipeline);
			command.setUniform(0, parameters);
			command.drawArrays(3);
			
			// your rendering code here
			
			return true;
		}
		
		// main run loop
		virtual bool run() {
			
			// run application
			if(window) return window.run([this]() {
				return render_window();
			});
			
			return true;
		}
		
	private:
		
		// helper function to create and initialize the Window
		virtual bool create_window() {
			
			// create Window
			window = Window(app.getPlatform(), app.getDevice());
			if(!window || !window.setSize(app.getWidth(), app.getHeight()) ||
				!window.create(String(window.getPlatformName()) + " @NAME@") || !window.setHidden(false)) {
				TS_LOG(Error, "Application::create_window(): can't create Window\n");
				return false;
			}
			
			// set Window callbacks
			window.setKeyboardPressedCallback([this](uint32_t key, uint32_t code) {
				if(key == Window::KeyF4 && window.getKeyboardKey(Window::KeyAlt)) window.stop();
				if(key == Window::KeyEsc) window.stop();
			});
			window.setCloseClickedCallback([&]() { window.stop(); });
			
			// create Device
			device = Device(window);
			if(!device) {
				TS_LOG(Error, "Application::create_window(): can't create Device\n");
				return false;
			}
			
			// create Target
			target = device.createTarget(window);
			if(!target) {
				TS_LOG(Error, "Application::create_window(): can't create Target\n");
				return false;
			}
			
			return true;
		}
		
		// helper function to create and load resources
		virtual bool create_resources() {
			
			// create pipeline
			pipeline = device.createPipeline();
			pipeline.setUniformMask(0, Shader::MaskFragment);
			pipeline.setColorFormat(window.getColorFormat());
			pipeline.setDepthFormat(window.getDepthFormat());
			if(!pipeline.loadShaderGLSL(Shader::TypeVertex, "main.shader", "VERTEX_SHADER=1")) {
				TS_LOG(Error, "Application::create_resources(): can't load Vertex shader\n");
				return false;
			}
			if(!pipeline.loadShaderGLSL(Shader::TypeFragment, "main.shader", "FRAGMENT_SHADER=1")) {
				TS_LOG(Error, "Application::create_resources(): can't load Fragment shader\n");
				return false;
			}
			if(!pipeline.create()) {
				TS_LOG(Error, "Application::create_resources(): can't create Pipeline\n");
				return false;
			}
			
			// your resource creation code here
			
			return true;
		}
		
		// helper function to render the Window
		virtual bool render_window() {
			
			// update events
			Window::update();
			
			// update application
			if(!update()) return false;
			
			// render window
			if(!window.render()) return false;
			
			// window target
			target.begin();
			{
				Command command = device.createCommand(target);
				
				// render application
				if(!render(command)) {
					target.end();
					return false;
				}
			}
			target.end();
			
			// present window
			if(!window.present()) return false;
			
			// check errors
			if(!device.check()) return false;
			
			return true;
		}
		
		App app;				// Application interface
		Window window;			// Window interface
		Device device;			// Device interface
		Target target;			// Target interface
		
		Pipeline pipeline;		// Pipeline interface
};

/*
 */
int32_t main(int32_t argc, char **argv) {
	
	// instantiate the application
	Application application(argc, argv);
	
	// create the application
	if(!application.create()) return 1;
	
	// run the application
	if(!application.run()) return 1;
	
	return 0;
}

/*
 */
#if _WIN32
	#include <system/TellusimWindows.h>
	TS_DECLARE_WINDOWS_HIGH_PERFORMANCE
#elif _ANDROID
	#include <system/TellusimAndroid.h>
	TS_DECLARE_ANDROID_NATIVE_ACTIVITY
#endif

File: tools/project/templates/rust/app_core_fragment.rs
// Tellusim Fragment Application Template

use tellusim::*;
use std::process::exit;

/*
 */
struct Application {
	
	app: App,				// Application interface
	window: Window,			// Window interface
	device: Device,			// Device interface
	target: Target,			// Target interface
	
	pipeline: Pipeline,		// Pipeline interface
}

impl Application {
	
	// constructor initializing the application with command-line arguments
	fn new() -> Application {
		Application {
			app: App::new(),
			window: Window::null(),
			device: Device::null(),
			target: Target::null(),
			pipeline: Pipeline::null(),
		}
	}
	
	// create function responsible for initialization
	fn create(&mut self) -> bool {
		
		// create app
		if !self.app.create() {
			ts_log!(Error, "Application::create(): can't create App\n");
			return false
		}
		
		// create window
		if !self.create_window() {
			return false
		}
		
		// create resources
		if !self.create_resources() {
			return false
		}
		
		// additional initialization code can go here
		
		true
	}
	
	// update function for logic updates
	fn update(&mut self) -> bool {
		
		// your update logic here
		
		true
	}
	
	// render function to handle the drawing commands
	fn render(&mut self, command: &mut Command) -> bool {
		
		// fragment uniform parameters
		#[repr(C)]
		#[derive(Default)]
		struct FragmentParameters {
			aspect: f32,
			time: f32,
		}
		let mut parameters = FragmentParameters::default();
		parameters.aspect = self.window.width() as f32 / self.window.height() as f32;
		parameters.time = time::seconds() as f32;
		
		// draw fullscreen triangle
		command.set_pipeline(&mut self.pipeline);
		command.set_uniform(0, &parameters);
		command.draw_arrays(3);
		
		// your rendering code here
		
		true
	}
	
	// main run loop
	fn run(&mut self) -> bool {
		
		// run application
		let mut window = self.window.copy_ptr();
		window.run(|| -> bool {
			self.render_window()
		});
		
		// finish context
		self.window.finish();
		
		true
	}
	
	// helper function to create and initialize the Window
	fn create_window(&mut self) -> bool {
		
		// create Window
		let mut window = Window::new_with_platform_index(self.app.platform(), self.app.device());
		if !window.is_valid_ptr() || !window.set_size(self.app.width(), self.app.height()) ||
			!window.create_with_title(&(window.platform_name() + &" Rust @NAME@")) || !window.set_hidden(false) {
			ts_log!(Error, "Application::create_window(): can't create Window\n");
			return false
		}
		
		// set Window callbacks
		window.set_keyboard_pressed_callback({
			let mut window = window.copy_ptr();
			move |key: u32, _code: u32| {
				if key == WindowKey::F4 && window.keyboard_key(WindowKey::Alt.value()) { window.stop() }
				if key == WindowKey::Esc { window.stop() }
			}
		});
		window.set_close_clicked_callback({
			let mut window = window.copy_ptr();
			move || { window.stop() }
		});
		
		// create Device
		let device = Device::new_with_window(&mut window);
		if !device.is_valid_ptr() {
			ts_log!(Error, "Application::create_window(): can't create Device\n");
			return false
		}
		
		// create Target
		let target = device.create_target_with_window(&mut window);
		if !target.is_valid_ptr() {
			ts_log!(Error, "Application::create_window(): can't create Target\n");
			return false
		}
		
		self.window = window;
		self.device = device;
		self.target = target;
		
		true
	}
	
	// helper function to create and load resources
	fn create_resources(&mut self) -> bool {
		
		// create pipeline
		let mut pipeline = self.device.create_pipeline();
		pipeline.set_uniform_mask(0, ShaderMask::Fragment);
		pipeline.set_color_format(0, self.window.color_format());
		pipeline.set_depth_format(self.window.depth_format());
		if !pipeline.load_shader_glsl(ShaderType::Vertex, "main.shader", "VERTEX_SHADER=1") {
			ts_log!(Error, "Application::create_resources(): can't load Vertex shader\n");
			return false;
		}
		if !pipeline.load_shader_glsl(ShaderType::Fragment, "main.shader", "FRAGMENT_SHADER=1") {
			ts_log!(Error, "Application::create_resources(): can't load Fragment shader\n");
			return false
		}
		if !pipeline.create() {
			ts_log!(Error, "Application::create_resources(): can't create Pipeline\n");
			return false
		}
		
		self.pipeline = pipeline;
		
		// your resource creation code here
		
		true
	}
	
	// helper function to render the Window
	fn render_window(&mut self) -> bool {
		
		// update events
		Window::update();
		
		// update application
		if !self.update() { return false }
		
		// render window
		if !self.window.render() { return false }
		
		// window target
		self.target.begin();
		{
			let mut command = self.device.create_command_with_target(&mut self.target);
			
			// render application
			if !self.render(&mut command) {
				self.target.end();
				return false
			}
		}
		self.target.end();
		
		// present window
		if !self.window.present() { return false }
		
		// check errors
		if !self.device.check() { return false }
		
		true
	}
}

/*
 */
fn main() {
	
	// instantiate the application
	let mut application = Application::new();
	
	// create the application
	if !application.create() { exit(1) }
	
	// run the application
	if !application.run() { exit(1) }
}
